appserver/data/data-core/src/main/java/fish/payara/jakarta/data/core/cdi/extension/RepositoryImpl.java

```diff
@@ -118,6 +125,92 @@ public Object invoke(Object proxy, Method method, Object[] args) throws Throwabl
         return objectToReturn;
     }
 
+    private void evaluateDataQuery(QueryData dataForQuery, Method method) {
+        if (dataForQuery.getDeclaredEntityClass() == null) {
+            Class<?> entityType = findEntityTypeInMethod(method);
+            if (entityType != null) {
+                dataForQuery.setDeclaredEntityClass(entityType);
+                return;
+            }
+            Class<?> declaringClass = method.getDeclaringClass();
+            Method[] allMethods = declaringClass.getMethods();
+            for (Method interfaceMethod : allMethods) {
+                if (interfaceMethod.equals(method)) {
+                    continue;
+                }
+                entityType = findEntityTypeInMethod(interfaceMethod);
+                if (entityType != null) {
+                    dataForQuery.setDeclaredEntityClass(entityType);
+                    return;
+                }
+            }
+            throw new MappingException(
+                    String.format("Could not determine primary entity type for repository method '%s' in %s. " +
+                                    "Either extend a repository interface with entity type parameters or " +
+                                    "ensure entity type is determinable from method signature.",
+                            method.getName(), declaringClass.getName())
+            );
+
+        }
+    }
+
+    private Class<?> findEntityTypeInMethod(Method method) {
```

Comment:
Could you move this method on a different location more generic? I think a possible place could be the class DataCommonOperationUtility and also make that method static